src/commons/crypto/signing.rs

```diff
@@ -36,84 +40,244 @@ use crate::{
+        }
+        .map_err(crypto::Error::signer)
+    }
+
+    pub fn destroy_key(&mut self, key_id: &KeyIdentifier) -> CryptoResult<()> {
+        match self {
+            SignerProvider::OpenSsl(signer) => signer.destroy_key(key_id),
+            #[cfg(feature = "hsm")]
+            SignerProvider::Dummy(signer) => signer.destroy_key(key_id),
+        }
+        .map_err(crypto::Error::key_error)
+    }
+
+    pub fn get_key_info(&self, key_id: &KeyIdentifier) -> CryptoResult<PublicKey> {
+        match self {
+            SignerProvider::OpenSsl(signer) => signer.get_key_info(key_id),
+            #[cfg(feature = "hsm")]
+            SignerProvider::Dummy(signer) => signer.get_key_info(key_id),
+        }
+        .map_err(crypto::Error::key_error)
+    }
+
+    pub fn random_serial(&self) -> CryptoResult<Serial> {
+        match self {
+            SignerProvider::OpenSsl(signer) => Serial::random(signer.deref()),
+            #[cfg(feature = "hsm")]
+            SignerProvider::Dummy(signer) => Serial::random(signer.deref()),
+        }
+        .map_err(crypto::Error::signer)
+    }
+
+    pub fn sign<D: AsRef<[u8]> + ?Sized>(
+        &self,
+        key_id: &KeyIdentifier,
+        sig_alg: SignatureAlgorithm,
+        data: &D,
+    ) -> CryptoResult<Signature> {
+        match self {
+            SignerProvider::OpenSsl(signer) => signer.sign(key_id, sig_alg, data),
+            #[cfg(feature = "hsm")]
+            SignerProvider::Dummy(signer) => signer.sign(key_id, sig_alg, data),
+        }
+        .map_err(crypto::Error::signing)
+    }
+
+    pub fn sign_one_off<D: AsRef<[u8]> + ?Sized>(
+        &self,
+        sig_alg: SignatureAlgorithm,
+        data: &D,
+    ) -> CryptoResult<(Signature, PublicKey)> {
+        match self {
+            SignerProvider::OpenSsl(signer) => signer.sign_one_off(sig_alg, data),
+            #[cfg(feature = "hsm")]
+            SignerProvider::Dummy(signer) => signer.sign_one_off(sig_alg, data),
+        }
+        .map_err(crypto::Error::signer)
+    }
+
+    pub fn sign_csr(&self, base_repo: &RepoInfo, name_space: &str, key: &KeyIdentifier) -> CryptoResult<Csr> {
+        fn func<T>(signer: &T, base_repo: &RepoInfo, name_space: &str, key: &KeyIdentifier) -> CryptoResult<Captured>
+        where
+            T: Signer<KeyId = KeyIdentifier>,
+        {
+            let pub_key = signer.get_key_info(key).map_err(crypto::Error::key_error)?;
+            Csr::construct(
+                signer.deref(),
+                key,
+                &base_repo.ca_repository(name_space).join(&[]).unwrap(), // force trailing slash
+                &base_repo.rpki_manifest(name_space, &pub_key.key_identifier()),
+                Some(&base_repo.rpki_notify()),
+            )
+            .map_err(crypto::Error::signing)
+        }
+
+        let enc = match self {
+            SignerProvider::OpenSsl(signer) => func(signer.deref(), base_repo, name_space, key),
```

Comment:
I am (generally) not fond of fn names like 'func'. Perhaps a more descriptive thing like 'sing_csr_with_signer'? Or.. is there a way to use the match to get the signer to use without the need for an inner fn?

Reply:
Oops, that should indeed not be called `func`.